osquery/core/database/database.h

```diff
@@ -0,0 +1,79 @@
+/**
+ *  Copyright (c) 2018-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#pragma once
+
+#include <cstdint>
+
+#include <osquery/error.h>
+#include <osquery/expected.h>
+#include <osquery/logger.h>
+
+namespace osquery {
+
+enum class DatabaseError {
+  // Unknown error, currently unused
+  Unknown = 1,
+  DatabaseIsNotOpen = 2,
+  DatabasePathDoesNotExists = 3,
+  FailToDestroyDB = 4,
+  FailToOpenDatabase = 5,
+  FailToReadData = 6,
+  FailToWriteData = 7,
+  KeyNotFound = 8,
+  DomainNotFound = 9,
+  // Corruption or other unrecoverable error after DB can't be longer used
+  // Database should be closed, destroyed and opened again
+  // If this error was received during data access, then aplication
+  // is likely to die soon
+  // See message and/or underlying error for details
+  Panic = 10,
+};
+
+class Database {
+public:
+  explicit Database(std::string name): name_(std::move(name)) {};
+  virtual ~Database() {};
+
+  const std::string& getName() const {
+    return name_;
+  }
+
+  virtual ExpectedSuccess<DatabaseError> open() = 0;
+  virtual ExpectedSuccess<DatabaseError> destroyDB() = 0;
+  virtual void close() = 0;
```

Comment:
Just a question, what is the difference between destroying and closing db?

Reply:
closing will close current connection to database and destroying will drop database.

Reply:
May I ask you to write it down above this methods as a docstring? Just to be sure that no-one miss it?